graphbolt/src/cuda/index_select_impl.cu

```diff
@@ -46,19 +107,41 @@ torch::Tensor UVAIndexSelectImpl_(torch::Tensor input, torch::Tensor index) {
                                       .dtype(input.dtype())
                                       .device(c10::DeviceType::CUDA));
   DType* input_ptr = input.data_ptr<DType>();
-  IdType* index_ptr = index.data_ptr<IdType>();
   DType* ret_ptr = ret.data_ptr<DType>();
+
+  // Sort the index to improve the memory access pattern.
+  torch::Tensor sorted_index, permutation;
+  std::tie(sorted_index, permutation) = torch::sort(index);
+  const IdType* index_sorted_ptr = sorted_index.data_ptr<IdType>();
+  const int64_t* permutation_ptr = permutation.data_ptr<int64_t>();
+
   cudaStream_t stream = 0;
-  dim3 block(512, 1);
-  // Find the smallest block size that can fit the feature_size.
-  while (static_cast<int64_t>(block.x) >= 2 * feature_size) {
-    block.x >>= 1;
-    block.y <<= 1;
+
+  if (feature_size == 1) {
+    const int num_threads = cuda::FindNumThreads(return_len);
+    const int num_blocks = (return_len + num_threads - 1) / num_threads;
+    IndexSelectSingleKernel<<<num_blocks, num_threads, 0, stream>>>(
+        input_ptr, input_len, index_sorted_ptr, return_len, ret_ptr,
+        permutation_ptr);
+  } else {
+    dim3 block(512, 1);
+    while (static_cast<int64_t>(block.x) >= 2 * feature_size) {
+      block.x >>= 1;
+      block.y <<= 1;
+    }
+    const dim3 grid((return_len + block.y - 1) / block.y);
+    if (feature_size * sizeof(DType) < 2 * GPU_CACHE_LINE_SIZE) {
```

Comment:
@mfbalin Could you explain why we choose `IndexSelectMultiKernel` in this case? From my understanding, `IndexSelectMultiKernelAligned` should be better than `IndexSelectMultiKernel` in any cases.

Reply:
Using fewer SMs will be important when we want to overlap the copies with other ongoing computations. I will revisit this code to minimize SM usage further.